Sources/NIOHTTP1/NIOHTTPClientUpgradeHandler.swift

```diff
@@ -0,0 +1,434 @@
+    ///     after the upgrade. It should include any handlers that are directly related to handling HTTP.
+    ///     At the very least this should include the `HTTPEncoder` and `HTTPDecoder`, but should also include
+    ///     any other handler that cannot tolerate receiving non-HTTP data.
+    /// - Parameter upgradeCompletionHandler: A closure that will be fired when HTTP upgrade is complete.
+    public init(upgraders: [NIOHTTPClientProtocolUpgrader],
+                httpHandlers: [RemovableChannelHandler],
+                upgradeCompletionHandler: @escaping (ChannelHandlerContext) -> Void) {
+
+        precondition(upgraders.count > 0, "A minimum of one protocol upgrader must be specified.")
+
+        self.upgraders = upgraders
+        self.httpHandlers = httpHandlers
+        self.upgradeCompletionHandler = upgradeCompletionHandler
+    }
+
+    public func write(context: ChannelHandlerContext, data: NIOAny, promise: EventLoopPromise<Void>?) {
+        
+        precondition(upgraders.count > 0, "A minimum of one protocol upgrader must be specified.")
+
+        switch self.upgradeState {
+        
+        case .requestRequired:
+            let updatedData = self.addHeadersToOutboundOut(data: data)
+            context.write(updatedData, promise: promise)
+
+        case .awaitingConfirmationResponse:
+            // Still have full http stack.
+            context.write(data, promise: promise)
+            
+        case .awaitingUpgrader, .upgraderReady, .upgrading:
+            preconditionFailure("We shouldn't be sending anything else until we get a response.")
+
+        case .upgradingAddingHandlers:
+            // These are most likely messages immidiately fired by a new protocol handler.
+            // As that is added last we can just forward them on.
+            context.write(data, promise: promise)
+            
+        case .upgradeComplete:
+            preconditionFailure("Upgrade complete and this handler should have been removed from the pipeline.")
+        case .upgradeFailed:
+            preconditionFailure("Upgrade failed and this handler should have been removed from the pipeline.")
+        }
+    }
+
+    private func addHeadersToOutboundOut(data: NIOAny) -> NIOAny {
+        
+        let interceptedOutgoingRequest = self.unwrapOutboundIn(data)
+        
+        if case .head(let requestHead) = interceptedOutgoingRequest {
+            
+            self.upgradeState = .awaitingConfirmationResponse
+            
+            let updatedHeader = self.addUpgradeHeaders(to: requestHead)
+            let completedHeader = self.addConnectionHeaders(to: updatedHeader)
+            return self.wrapOutboundOut(.head(completedHeader))
+        }
+        
+        return data
+    }
+    
+    private func addUpgradeHeaders(to requestHead: HTTPRequestHead) -> HTTPRequestHead {
+        
+        var head = requestHead
+        
+        head.headers.add(name: "Connection", value: "upgrade")
```

Comment:
Maybe move this `headers.add` call to `addConnectionHeaders` and then build the entire header string and add it in a single go? That will avoid producing extra overhead in the form of repeating the `Connection: ` string on the wire. We can then also move the `addConnectionHeaders` call to before the `addUpgradeHeaders` call to make things easier for any proxies that may be in the way.